pkg/cgroups/cgroupsv2/cgroup_manager.go

```diff
@@ -0,0 +1,127 @@
+package cgroupsv2
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"slices"
+
+	jobv1 "github.com/kralicky/jobserver/pkg/apis/job/v1"
+)
+
+const (
+	hierarchyRootPath = "/sys/fs/cgroup"
+	jobserverCgroup   = "kralicky-jobserver"
+)
+
+var (
+	requiredControllers = []string{"cpu", "memory", "io"}
+)
+
+type cgroupManager struct {
+	path string
+}
+
+func newCgroupManager() (*cgroupManager, error) {
+	if ok, err := requiredControllersEnabled(filepath.Join(hierarchyRootPath, "cgroup.controllers")); !ok {
+		return nil, err
+	}
+	if ok, err := requiredControllersEnabled(filepath.Join(hierarchyRootPath, "cgroup.subtree_control")); !ok {
+		return nil, err
+	}
+	// create the jobserver cgroup if it doesn't exist
+	jobserverCgroup := filepath.Join(hierarchyRootPath, jobserverCgroup)
+	if _, err := os.Stat(jobserverCgroup); os.IsNotExist(err) {
+		if err := os.Mkdir(jobserverCgroup, 0755); err != nil {
+			return nil, fmt.Errorf("failed to create jobserver cgroup: %w", err)
+		}
+		slog.Info("created jobserver cgroup", "path", jobserverCgroup)
+	}
+
+	// ensure the required controllers are enabled
+	if err := enableRequiredControllers(filepath.Join(jobserverCgroup, "cgroup.subtree_control")); err != nil {
+		return nil, fmt.Errorf("failed to enable cgroup subtree controllers: %w", err)
+	}
+	slog.Info("initialized jobserver cgroup", "path", jobserverCgroup)
+	return &cgroupManager{path: jobserverCgroup}, nil
+}
+
+func (m *cgroupManager) CreateCgroupWithLimits(id string, limits *jobv1.ResourceLimits) (string, error) {
+	// create a new cgroup for the job
+	path := filepath.Join(m.path, id)
+	if err := os.Mkdir(path, 0755); err != nil {
+		return "", fmt.Errorf("failed to create cgroup %s: %w", path, err)
+	}
+	slog.Info("created cgroup", "path", path, "job", id)
+
+	if limits == nil {
+		return path, nil
+	}
+
+	// set all present limits
+	if limits.Cpu != nil {
+		if err := writeCpuMaxQuota(path, mcpusToCfsQuota(limits.GetCpu())); err != nil {
+			return "", fmt.Errorf("failed to set cpu.max: %w", err)
+		}
+	}
+	if limits.Memory != nil {
+		if limits.Memory.SoftLimit != nil {
+			if err := writeMemoryHigh(path, limits.Memory.GetSoftLimit()); err != nil {
```

Comment:
Since we're checking pointers...

```suggestion
			if err := writeMemoryHigh(path, *limits.Memory.SoftLimit); err != nil {
```